### PR TITLE
Flatten error by removing ActionError

### LINKED ISSUE #119: Flatten out underlyingErrors
### Currently 

The current implementation of `Action` doesn't allow you to receive the `Swift.Error` when `.execution` is called. But we get an `ActionError` type which is not really useful because you always have to switch within it to the `underlyingError`.

### Proposal 

I think it would be nice for the framework to have something like `var underlyingErrors: Observable<Swift.Error> ` which will only emit when `.execution` actually has an error

### PR DESCRIPTION
**Pull request by dangthaison91**
This replaces PR for #138 and will resolve #119 (original by @bobgodwinx )

After referring Action's implementation from ReactiveSwift, I decide to add disabledErrors: Observable<Error> and remove ActionError. This will make breaking change but will help end users avoid misunderstand between .errors/.underlyingErrors and .executionErrors from the original proposal. I add ActionDisabledError for better debugging.

cc @ashfurrow @freak4pc

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -2,12 +2,26 @@ Changelog
...
+Master: Version table
+-------------
+
+| Swift version | RxSwift version | Action version |
+| ------------- | --------------- | -------------- |
+| Swift 3.0     | v3.2.*             | v2.2.0            |
+| Swift 3.2     | v3.6.*             | v3.2.0            |
+| **Swift 4**   | **v4.0.0**      | **v3.4.0**     |
+| Swift 4.1    | **v4.2.0**      | **v3.6.0**     |
+| Swift 4.2    | **v4.3.1**      |   **v4.0**     |
```
Comment (freak4pc) [inline Changelog.md]: Can we align the markdown? I know its rendered right but as a person who likes to read raw markdown it bugs me :P

File: Sources/Action/Action.swift [hunk 2]
```diff
@@ -30,7 +34,11 @@ public final class Action<Input, Element> {
 
     /// Errors aggrevated from invocations of execute().
     /// Delivered on whatever scheduler they were sent from.
-    public let errors: Observable<ActionError>
+    public let errors: Observable<Error>
+    
+    /// Errors when Action cannot execute due to disabled.
```
Comment (freak4pc) [inline Sources/Action/Action.swift]: What does this actually means? "due to being disabled", you meant?

File: Sources/Action/Action.swift [hunk 3]
```diff
@@ -30,7 +34,11 @@ public final class Action<Input, Element> {
...
+    /// Delivered on whatever scheduler they were sent from.
```
Comment (freak4pc) [inline Sources/Action/Action.swift]:
```suggestion
    /// Delivered on the scheduler they originated from.
```

**Comment by freak4pc** [comment]
If flattening is what you want, why not just expose a property ? 

```swift
public enum ActionError: Error {
    case notEnabled
    case underlyingError(Error)
    
    var underlyingError: Error? {
        guard case .underlyingError(let error) = self else { return nil }
        return error
    }
}
```

In my eyes, flattening should mean we still have access to both error types, not just "squish" them into one error and lose granularity.

**Comment by bobgodwinx** [comment]
@freak4pc using a `struct` is totally fine. I think the real `Error` should be left out of disabling and enabling. The code example you wrote was actually how I imagined it. But it tends to be confusing in the long run when `Error` emits and the work factory doesn't `Complete`. So pulling them apart is actually a good. Don't you think so?

File: Changelog.md (hunk 1 above)
Comment (dangthaison91) [reply Changelog.md]: Yeah, I will align them, just want to keep the PR is clean by keeping the original code.

File: Sources/Action/Action.swift (hunk 2 above)
Comment (dangthaison91) [reply Sources/Action/Action.swift]: Your action will be disabled by `enabledIf: Observable<Bool>` and `executing: Observable<Bool>`

**Comment by dangthaison91** [comment]
@freak4pc I use Action so much and all the time I only need to care about `underlyingError`.  But sometimes, for new user `disabledErrors` will help them to debug and give them ideas about `workFactory` must `completed` to be enabled again.

This also makes align API with Action in `ReactiveSwift`.

File: Sources/Action/Action.swift (hunk 3 above)
Comment (dangthaison91) [reply Sources/Action/Action.swift]: I just want to keep original code/comment that I don't touch

**Comment by ashfurrow** [comment]
Hmm, I'm not really sure now. I initially thought this was a good idea (see my comment in #119) but I like @freak4pc's solution of the computed property. This would avoid a breaking change and avoid dividing errors into two properties (underlying errors and disabled errors). What do you think, @dangthaison91? Sorry for flip-flopping!

**Comment by dangthaison91** [comment]
@ashfurrow @freak4pc 's suggestion is exactly what we are doing right now but it is a little bit confusing for our newcomers when they use Action. They can be confused by `.underlyingError` and original `errors`. We just want to simplify the error API.

OTW, We can help people understand Action's behavior better by adding docs for handling errors.

**Comment by ashfurrow** [comment]
@dangthaison91 that's a good point. I think that @freak4pc's solution would be most complete with additional docs 👍

**Comment by freak4pc** [comment]
Agree IRT to docs, they're very lacking ATM in general (IMO) and could use some love. I really want to have some time to do this but TBH, it's probably better for someone else who uses it daily to write it :) 

IRT to error flattening I still believe using a simple computed properly provides a singular "flattened" error without having to switch over anything, while still leaving the old behavior in tact and will act as a purely additive change - win win situation, in my book :)

**Comment by dangthaison91** [comment]
@freak4pc Should we ask people on Community channel how willing they are if we make a breaking change on Error API?

**Comment by freak4pc** [comment]
I think you'll need to get more users of Action who are actually convinced changing this is the right thing. It's not just about a breaking change, it's really about the fact I don't think this change provides so much value, TBH.

**Comment by bobgodwinx** [comment]
Hi @all This is the way I see it. `enabled ` and `disabled` are things that we trigger when the work factory is `executing`. It's not an error. So I think real `Errors` should be treated as such in a different channel as a separate property. I might be wrong but it works for me in my app perfectly.

**Comment by freak4pc** [comment]
That's actually an interesting point. OK, I sort of see what you're saying and I also re-read some of the code again. 

I imagine the case for `.notEnabled` isn't one you often want to handle together with errors emitted from the `workFactory` - correct ? 

In general, would you say a `notEnabled` error actually 100% of the time means a programmer has made an error? Or could it be a valid state? 

I think my main points of this are :

1. If this isn't an Error, don't make it a `Swift.Error` to begin with. 
2. If we're making breaking changes, let's think about the nicest way and not just a "middle" way. 

I want to talk about this a bit more to make sure we're all aligned, if we're going down that path :)

File: Changelog.md [hunk 4]
```diff
@@ -2,12 +2,26 @@ Changelog
...
 --------------
 
-4.0 
+4.0
 --------------------
 
+- Flatten error API [#158](https://github.com/RxSwiftCommunity/Action/pull/158)
+   - Removed  `ActionError`
+   - Added `errors: Observable<Error>` as underlying errors observable
+   - Added  `disabledErrors: Observable<AationDisabledError>` for better debugging when workFactory not send `onCompleted()`
```
[the remark this answers is not in the document]
Comment (khoi) [reply Changelog.md]:
```suggestion
   - Added  `disabledErrors: Observable<ActionDisabledError>` for better debugging when workFactory not send `onCompleted()`
```

**Comment by bobgodwinx** [comment]
@All I think @freak4pc got my point correct. We should find a better way to handle [execution state](https://github.com/RxSwiftCommunity/Action/blob/master/Sources/Action/Action.swift#L87) For now I don't see a better way and if we treat it as error then the old code is just fine. Please  pour in ideas.

**Comment by dangthaison91** [comment]
For now, we already had `.isEnabled` and `.isExecuting` as execution state. So, if we still want to continue with this PR, should we remove `disabledError` completely? We must add docs to help users aware of Action behavior with `completed`.

**Comment by bobgodwinx** [comment]
I am 50/50 for the documentation vs code. I believe in an `API` able to express it's self in what it's doing. The problem is that `Action` tends to throw the `.notEnabled`  often and it's confusing. So this is why separating this into another channel might be good even though it's not the best idea. However let's think through all options.

**Comment by freak4pc** [comment]
Execution state is fine, but what you want to convey here is that "an input was received while the Action is disabled, you should check your WorkFactory stream to make sure it completed properly", correct? 

In that case I'd say it _is_ an error, and I'd be even more for having it as part of a simple computed property like I mentioned above - making it the developer's choice to handle that or not. They can simply ignore the `actionDisabled` error (or whatever you'd call it). This way we're still keeping backwards compatibility and it would make 100% sense for most use cases IMHO. 

One more thing we can do here is adding a debug-only fatalError (much like we do in [RxSwift-core](https://github.com/ReactiveX/RxSwift/blob/53cd723d40d05177e790c8c34c36cec7092a6106/RxSwift/Rx.swift#L45)), or even less aggressive - a debug-only runtime warning.

We can also possibly make the computed properties a bit more obvious / self-documenting: 

```swift
var factoryError: Swift.Error? { // ... only underlying error } 
var isNotEnabled: Bool { // basic true/false without the error itself as it doesn't matter }
```

Thought?

**Comment by ashfurrow** [comment]
I would prefer @freak4pc's solution above if for no other reason than I don't want the API to encourage developers to ignore errors. In my opinion, calling `execute()` on an Action while it's already executing (and doesn't support concurrent execution) should produce an error. Users can ignore it if they want, but it shouldn't be separated into its own error property to make it easy to ignore, if that makes sense.

**Comment by freak4pc** [comment]
I think @ashfurrow's note makes sense. We shouldn't take steps to make this error easier to ignore as it might be hiding bigger issues.

**Comment by freak4pc** [comment]
> I think adding fatalError like RxSwift won't help cause action can emit notEnabled very often. For example, I used Action to implement loadmore/pagination API while don't want to block user interaction.

Agree. I still think having a computed property + fixing the documentation would be the best path moving forward.

**Comment by Blackening999** [comment]
How can one retrieve the underlying error?

**Comment by freak4pc** [comment]
Are you still interested in pursing this PR @dangthaison91 ?

**Comment by freak4pc** [comment]
@dangthaison91 @bobgodwinx @fpillet Anyone interested in pursuing this approach? Otherwise I think it's worth closing as it's been open for a while with no progress.

**Comment by dangthaison91** [comment]
IMHO, user can easily create a computed property so not sure how can I help to improve this @freak4pc

**Comment by bobgodwinx** [comment]
I think this issue is still a good one. Let me see what I can do in the coming days. @dangthaison91  + @freak4pc

**Pull request closed without merging**